sw/device/lib/dif/dif_gpio.h

```diff
@@ -0,0 +1,314 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef SW_DEVICE_LIB_DIF_GPIO_H
+#define SW_DEVICE_LIB_DIF_GPIO_H
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Configuration for initializing a GPIO device.
+ */
+typedef struct dif_gpio_config { mmio_region_t base_addr; } dif_gpio_config_t;
+
+/**
+ * Internal state of a GPIO device.
+ *
+ * Instances of this struct must be initialized by |dif_gpio_init()| before
+ * being passed to other functions.
+ */
+typedef struct dif_gpio { mmio_region_t base_addr; } dif_gpio_t;
+
+/**
+ * Generic return codes for the functions in this library.
+ */
+typedef enum dif_gpio_result {
+  kDifGpioResultOK,
+  kDifGpioResultInvalidArgument,
+} dif_gpio_result_t;
+
+/**
+ * Set of allowed interrupt trigger configurations.
+ */
+typedef enum dif_gpio_interrupt_trigger {
+  kDifGpioInterruptTriggerDisableAll,
+  kDifGpioInterruptTriggerEnableRisingEdge,
+  kDifGpioInterruptTriggerEnableFallingEdge,
+  kDifGpioInterruptTriggerEnableLevelLow,
+  kDifGpioInterruptTriggerEnableLevelHigh,
+  kDifGpioInterruptTriggerEnableRisingEdgeAndFallingEdge,
```

Comment:
The only comment I would have is that the overly long constants like `kDifGpioInterruptTriggerEnableRisingEdgeAndFallingEdge` are a bit awkward. I had the same issue when was working on PLIC, please consider something like:
```
typedef enum dif_gpio_irq_trigger {
...
￼  kDifGpioIrqTriggerRisingEdgeFallingEdge,
...
}

```
And I think `kDifGpioInterruptTriggerDisableAll` probably could be a separate routine. Would be nice to have another opinion on this.

Reply:
Or even:
```
typedef enum dif_gpio_irq_trigger {
...
kDifGpioIrqTriggerEdgeRisingFalling,
...
kDifGpioIrqTriggerEdgeFallingLevelHigh,
...
}
```

Reply:
Done! Thanks for pointing this out. I also replaced `_interrupt_` in other functions with `_irq_` for the sake of consistency.